presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/json/JsonRowEncoder.java

```diff
@@ -40,32 +56,71 @@
     {
         super(session, columnHandles);
 
+        ImmutableList.Builder<JsonDateTimeFormatter> formatFunctions = ImmutableList.builder();
         for (EncoderColumnHandle columnHandle : this.columnHandles) {
             checkArgument(isSupportedType(columnHandle.getType()), "Unsupported column type '%s' for column '%s'", columnHandle.getType(), columnHandle.getName());
-            checkArgument(columnHandle.getFormatHint() == null, "Unexpected format hint '%s' defined for column '%s'", columnHandle.getFormatHint(), columnHandle.getName());
-            checkArgument(columnHandle.getDataFormat() == null, "Unexpected data format '%s' defined for column '%s'", columnHandle.getDataFormat(), columnHandle.getName());
+
+            if (isSupportedTemporalType(columnHandle.getType())) {
+                checkArgument(columnHandle.getDataFormat() != null, "Unsupported or no dataFormat '%s' defined for temporal column '%s'", columnHandle.getDataFormat(), columnHandle.getName());
+                FormatType dataFormat = parseDataFormat(columnHandle.getDataFormat(), columnHandle.getName());
+                checkArgument(dataFormat.isSupportedType(columnHandle.getType()), "Unsupported column type '%s' for column '%s'", columnHandle.getType(), columnHandle.getName());
+
+                if (dataFormat == FormatType.CUSTOM_DATE_TIME) {
+                    checkArgument(columnHandle.getFormatHint() != null, "No format hint defined for column '%s'", columnHandle.getName());
+                }
+                else {
+                    checkArgument(columnHandle.getFormatHint() == null, "Unexpected format hint '%s' defined for column '%s'", columnHandle.getFormatHint(), columnHandle.getName());
+                }
+
+                formatFunctions.add(dataFormat.getFormatter(Optional.ofNullable(columnHandle.getFormatHint())));
+            }
+            else {
+                checkArgument(columnHandle.getFormatHint() == null, "Unexpected format hint '%s' defined for column '%s'", columnHandle.getFormatHint(), columnHandle.getName());
+                checkArgument(columnHandle.getDataFormat() == null, "Unexpected data format '%s' defined for column '%s'", columnHandle.getDataFormat(), columnHandle.getName());
+                formatFunctions.add(new UnimplementedJsonDateTimeFormatter());
+            }
         }
 
+        this.dateTimeFormatters = formatFunctions.build();
         this.objectMapper = requireNonNull(objectMapper, "objectMapper is null");
         this.node = objectMapper.createObjectNode();
     }
 
     private boolean isSupportedType(Type type)
     {
         return isVarcharType(type) ||
-                PRIMITIVE_SUPPORTED_TYPES.contains(type);
+                SUPPORTED_PRIMITIVE_TYPES.contains(type) ||
+                isSupportedTemporalType(type);
+    }
+
+    private boolean isSupportedTemporalType(Type type)
+    {
+        return SUPPORTED_TEMPORAL_TYPES.contains(type);
```

Comment:
Do we support only types with `precision == 3` intentionally?
To support different precisions you need to check using `type instanceof TimeType` etc.

Reply:
Yes that is intentional to mirror the decoder for now. Updating the decoder to use variable precision time/timestamps is going to be a bigger change so I wanted to constrain the encoder to default precision 3 for now.